lib/shape.ml

```diff
@@ -60,6 +51,72 @@ let line ?x1 ?y1 x2 y2 =
   | Some x, Some y -> Line { a = { x; y }; b = { x = x2; y = y2 } }
   | _ -> Line { a = { x = 0; y = 0 }; b = { x = x2; y = y2 } }
 
+let polygon lst_points = Polygon lst_points
+
+let rec take n lst =
+  match (n, lst) with
+  | 0, _ -> ([], lst)
+  | _, [] -> ([], [])
+  | n, x :: xs ->
+      let taken, rest = take (n - 1) xs in
+      (x :: taken, rest)
+
+let rec partition n step lst =
+  match lst with
+  | [] -> []
+  | _ ->
+      let taken, _ = take n lst in
+      if List.length taken = n then taken :: partition n step (List.tl lst)
+      else []
+
+let render_polygon polygon =
+  let denorm_points = List.map denormalize polygon in
+  let point_to_tuples =
+    List.map (fun { x : int; y : int } -> (x, y)) denorm_points
+  in
+  let partitioned =
+    partition 2 1 (point_to_tuples @ [ List.hd point_to_tuples ])
+  in
+  let array_of_points = Array.of_list (List.flatten partitioned) in
```

Comment:
why do we need to do this?

i think it should just work with

```ocaml
Array.of_list (List.map (fun { x: int; y: int} -> (x, y)))
```

Reply:
We need to do this to properly close the polygon, without this a face would be left open. This takes the list of points, for example [p1; p2; p3], and transforms it into [[p1; p2]; [p2; p3]; [p3; p1]] after which it is flattened

Reply:
Besides `draw_poly_line`, there is another function `draw_poly` in the graphics library that will automatically do this closing.